node/cli/src/service.rs

```diff
@@ -63,6 +66,60 @@ impl<F> Default for NodeConfig<F> where F: substrate_service::ServiceFactory {
 	}
 }
 
+/// A provider of current authority key.
+pub struct KeyProvider {
+	client: Arc<FullClient<Factory>>,
+	pub aura_store: keystore::LocalStore<AuraPair>,
+	pub grandpa_store: keystore::LocalStore<GrandpaPair>,
+}
+
+impl Clone for KeyProvider {
+	fn clone(&self) -> Self {
+		Self {
+			client: self.client.clone(),
+			aura_store: self.aura_store.clone(),
+			grandpa_store: self.grandpa_store.clone(),
+		}
+	}
+}
+
+impl offchain::AuthorityKeyProvider for KeyProvider {
+	fn authority_key<TPair: Pair>(&self) -> Option<TPair> {
+		if TPair::KEY_TYPE == <AuraPair as TypedKey>::KEY_TYPE {
```

Comment:
What if `AuraPair::KEY_TYPE == GrandpaPair::KEY_TYPE`? We will use a wrong `runtime_api` it seems.
If that's a case we should probably call both apis and just find the first key that works?

Reply:
if the key isn't found in the aura authorities then we move on to the grandpa_authorities.

Reply:
Oh, that's good then, I didn't notice. I like your idea splitting the API into two separate methods (one for grandpa, one for authority_key), but we can address that in a separate PR.